src/index.js

```diff
@@ -0,0 +1,234 @@
+  **/
+  unbindAdapter(adapterName) {
+    assert(adapterName);
+    if (this.adapters.has(adapterName)) {
+      delete instance[adapterName];
+      this.adapters.delete(adapterName);
+    }
+  }
+}
+
+/**
+ * Configure and return a chain of processing a log message. It should be required on module launch.
+ * It sets up loggers, notifiers for bugs, and exposes the internal providers, abstracted.
+ * If no parameter is provided, it will return already configured providers
+ * @param  {Object}  [config] Configuration object required by the internal providers
+ * @param  {Array}   [config.chainLinks] Array of instances of chainLinks
+ * @param  {Object}  [config.adapters] Object with adapters to be exposed
+ * @param  {boolean} [config.CONSOLE_LOGGING] Switches on/off the console logging chain
+ * @param  {boolean} [config.MIN_LOG_LEVEL] Minimal log level for a message to be processed
+ * @param  {boolean} [config.MIN_LOG_LEVEL_CONSOLE] Minimal log level for a message to be processed by console chain link (more prior over MIN_LOG_LEVEL)
+ * @param  {boolean} [config.DEFAULT_LOG_LEVEL] Fall back if a message does not have one
+ * Note that this module can also be configured with environment variables.
+ * Environment variables have higher priority over a settings object
+ * @return {Object}  The configured providers
+ */
+module.exports = (config) => {
+  if (!config) {
+    if (instance) return instance;
+    console.warn('Logtify should be initilised before used without config.');
+  }
+  const settings = Object.assign({}, config);
+
+  // presets
+  if (Array.isArray(settings.presets)) {
+    // a whole settings object is passed to initialize the chain links (if wrapped into a function)
+    const presetConfigs = preset(settings);
+    // merging other preset-given configs
+    Object.assign(settings, presetConfigs);
+  }
+
+  // if any have been bufferized
+  let adapters = buffer.adapters;
+  const chainLinks = buffer.chainLinks;
+
+  const chain = new LoggerChain(settings);
+  instance = { chain };
+
+  // adding default Console chain link
+  const chainLinkIndex = chain.push(new ConsoleChainLink(settings));
+  chain.bindAdapter('logger', new Winston(chain, chainLinkIndex));
+
+  // Custom chainLinks
+  if (Array.isArray(chainLinks)) {
+    for (const CustomChainLink of chainLinks) {
+      // if constructor passed in the array
+      if (typeof CustomChainLink === 'function') {
+        chain.push(new CustomChainLink(settings));
+        // if a pre-configured chainLink with some link-specific settings
+      } else if (CustomChainLink !== null && typeof CustomChainLink === 'object') {
+        const chainLinkConfig = Object.assign({}, CustomChainLink.config, settings);
+        const ChainLinkClass = CustomChainLink.class;
+        chain.push(new ChainLinkClass(chainLinkConfig));
+        // if a pre-configured object also exposes adapter
+        if (CustomChainLink.adapter !== null && typeof CustomChainLink.adapter === 'object') {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.dialonce.io/coding_rules#rule_key=javascript%3ANestedIfDepth)